apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,265 @@
+    if (copyDescriptors.length === 0) {
+      return [0, 0];
+    }
+
+    let copyCount: number = 0;
+    let hardlinkCount: number = 0;
+    await Async.forEachLimitAsync(
+      copyDescriptors,
+      MAX_PARALLELISM,
+      async (copyDescriptor: ICopyFileDescriptor) => {
+        if (copyDescriptor.hardlink) {
+          // Hardlink doesn't allow passing in overwrite param, so delete ourselves
+          try {
+            await FileSystem.deleteFileAsync(copyDescriptor.destinationFilePath, { throwIfNotExists: true });
+          } catch (e) {
+            if (!FileSystem.isFileDoesNotExistError(e)) {
+              throw e;
+            }
+            // Since the file doesn't exist, the parent folder may also not exist
+            await FileSystem.ensureFolderAsync(path.dirname(copyDescriptor.destinationFilePath));
+          }
+
+          await FileSystem.createHardLinkAsync({
+            linkTargetPath: copyDescriptor.sourceFilePath,
+            newLinkPath: copyDescriptor.destinationFilePath
+          });
+          hardlinkCount++;
+        } else {
+          // If it's a copy, simply call the copy function
+          await FileSystem.copyFileAsync({
+            sourcePath: copyDescriptor.sourceFilePath,
+            destinationPath: copyDescriptor.destinationFilePath,
+            alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+          });
+          copyCount++;
+        }
+      }
+    );
+
+    return [copyCount, hardlinkCount];
+  }
+
+  private async _getCopyFileDescriptorsAsync(
+    buildFolder: string,
+    copyConfigurations: IExtendedSharedCopyConfiguration[]
+  ): Promise<ICopyFileDescriptor[]> {
+    // Create a map to deduplicate and prevent double-writes
+    // resolvedDestinationFilePath -> [resolvedSourceFilePath, hardlink]
+    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+
+    for (const copyConfiguration of copyConfigurations) {
+      // Resolve the source folder path which is where the glob will be run from
+      const resolvedSourceFolderPath: string = path.resolve(buildFolder, copyConfiguration.sourceFolder);
+
+      // Glob extensions with a specific glob to increase perf
+      let sourceFileRelativePaths: Set<string>;
+      if (copyConfiguration.fileExtensions?.length) {
+        const escapedExtensions: string[] = globEscape(copyConfiguration.fileExtensions);
+        const pattern: string = `**/*+(${escapedExtensions.join('|')})`;
+        sourceFileRelativePaths = await this._expandGlobPatternAsync(
+          resolvedSourceFolderPath,
+          pattern,
+          copyConfiguration.excludeGlobs
+        );
+      } else {
+        sourceFileRelativePaths = new Set<string>();
+      }
+
+      // Now include the other glob as well
+      for (const include of copyConfiguration.includeGlobs || []) {
+        const explicitlyIncludedPaths: Set<string> = await this._expandGlobPatternAsync(
```

Comment:
It'd be cool if we could somehow parse these glob patterns to compiled regexes or something and then try to match those against an explicit crawling of the filesystem.

Reply:
My thought is that if compiled regexes are faster, this would be the approach fast-glob would take internally. Either that, or at the very least it would provide the option to do this. `micromatch`, which `fast-glob` relies on, has the ability to convert a glob into a regex, and I think fast-glob would try to take advantage of it given the speed-focused mentality of the lib. I could very easily be wrong here though.